ws/src/session.rs

```diff
@@ -72,6 +77,61 @@ impl From<Option<ws::Response>> for MiddlewareAction {
 	}
 }
 
+// the slab is only inserted into when live.
+type TaskSlab = Mutex<Slab<Option<oneshot::Sender<()>>>>;
+
+// future for checking session liveness.
+// this returns `NotReady` until the session it corresponds to is dropped.
+struct LivenessPoll {
+	task_slab: Arc<TaskSlab>,
+	slab_handle: usize,
+	rx: oneshot::Receiver<()>,
+}
+
+impl LivenessPoll {
+	fn create(task_slab: Arc<TaskSlab>) -> Self {
+		const INITIAL_SIZE: usize = 4;
+
+		let (index, rx) = {
+			let mut task_slab = task_slab.lock();
+			if !task_slab.has_available() {
+				// grow the size if necessary.
+				// we don't expect this to get so big as to overflow.
+				let reserve = ::std::cmp::max(task_slab.capacity(), INITIAL_SIZE);
```

Comment:
yep, it'll grow up to `next_power_of_two(maximum amount of live requests in session)`